Implement imperative invoker changes to popover APIs

lukewarlow opened the pull request:
#### 899ca0cd494c188aeb6bf1566f35e7b29bab8869
<pre>
Implement imperative invoker changes to popover APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=282337">https://bugs.webkit.org/show_bug.cgi?id=282337</a>

Reviewed by Tim Nguyen.

Update showPopover() and togglePopover() to be able to set
a popover invoker relationship imperatively.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/imperative-invokers-expected.txt:
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::invoker const):
(WebCore::PopoverData::setInvoker):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::togglePopover):
(WebCore::HTMLElement::handleCommandInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const):
* Source/WebCore/page/FocusController.cpp:
(WebCore::invokerForOpenPopover):

Canonical link: <a href="https://commits.webkit.org/287624@main">https://commits.webkit.org/287624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86acf079c1572e962b051fcaf4e108362d3bfa59

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80264 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33678 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84781 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31240 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82375 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68327 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7566 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62751 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20560 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73096 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43057 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50134 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27251 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29701 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71265 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27765 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86214 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7484 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5308 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71029 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7659 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68936 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70272 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17503 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14260 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13205 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7448 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12965 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7287 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10807 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9092 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1217,14 +1222,29 @@ ExceptionOr<void> HTMLElement::hidePopover()
+ExceptionOr<bool> HTMLElement::togglePopover(std::optional<std::variant<WebCore::HTMLElement::TogglePopoverOptions, bool>> value)
 {
+    std::optional<bool> force;
+    HTMLElement* invoker = nullptr;
```
```suggestion
    HTMLElement* invoker;
```

lukewarlow replied:
That change causes a runtime crash, I'm guessing because I'm doing &invoker and it's undefined?

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1217,14 +1222,29 @@ ExceptionOr<void> HTMLElement::hidePopover()
 }
 
-ExceptionOr<bool> HTMLElement::togglePopover(std::optional<bool> force)
+ExceptionOr<bool> HTMLElement::togglePopover(std::optional<std::variant<WebCore::HTMLElement::TogglePopoverOptions, bool>> value)
```
Can we name the argument `options` to match the spec ? We should be able to shadow `options` below, but `parameter` or something similar might be ok too if you want to avoid shadowing.

 `value` isn't super descriptive.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1217,14 +1222,29 @@ ExceptionOr<void> HTMLElement::hidePopover()
+                force = options.force;
+                invoker = options.source.get();
+            },
+            [&force](bool forceValue) {
```
I'd name this parameter `value`

nt1m reviewed (approved):
Thanks!

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1217,14 +1222,29 @@ ExceptionOr<void> HTMLElement::hidePopover()
+                force = options.force;
+                invoker = options.source.get();
+            },
+            [&force](bool value) {
```
```suggestion
            [&](TogglePopoverOptions options) {
                force = options.force;
                invoker = options.source.get();
            },
            [&](bool value) {
```
This is slightly more elegant if it works.

lukewarlow committed:
Implement imperative invoker changes to popover APIs
https://bugs.webkit.org/show_bug.cgi?id=282337

Reviewed by Tim Nguyen.

Update showPopover() and togglePopover() to be able to set
a popover invoker relationship imperatively.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/imperative-invokers-expected.txt:
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::invoker const):
(WebCore::PopoverData::setInvoker):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::togglePopover):
(WebCore::HTMLElement::handleCommandInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const):
* Source/WebCore/page/FocusController.cpp:
(WebCore::invokerForOpenPopover):

Canonical link: https://commits.webkit.org/287624@main
